library/src/main/java/com/novoda/downloadmanager/ExternalFilePersistence.java

```diff
@@ -47,28 +47,41 @@ public FilePersistenceResult create(FileName fileName, FileSize fileSize) {
             return FilePersistenceResult.newInstance(Status.ERROR_INSUFFICIENT_SPACE);
         }
 
-        String absolutePath = new File(externalFileDir, fileName.name()).getAbsolutePath();
-        FilePath filePath = FilePathCreator.create(absolutePath);
-        return create(filePath);
+        String absolutePath = new File(externalFileDir, filePath.path()).getAbsolutePath();
+        FilePath absoluteFilePath = FilePathCreator.create(absolutePath);
+        return create(absoluteFilePath);
     }
 
-    @Override
-    public FilePersistenceResult create(FilePath filePath) {
-        if (filePath.isUnknown()) {
-            return FilePersistenceResult.newInstance(Status.ERROR_OPENING_FILE, filePath);
+    private FilePersistenceResult create(FilePath absoluteFilePath) {
+        if (absoluteFilePath.isUnknown()) {
+            return FilePersistenceResult.newInstance(Status.ERROR_OPENING_FILE, absoluteFilePath);
         }
 
-        String absolutePath = filePath.path();
-
         try {
-            file = new File(absolutePath);
-            fileOutputStream = new FileOutputStream(absolutePath, APPEND);
+            file = new File(absoluteFilePath.path());
+            boolean parentDirectoriesCreated = createParentDirectoriesIfDoesNotExist(file);
+
+            if (!parentDirectoriesCreated) {
+                return FilePersistenceResult.newInstance(Status.ERROR_OPENING_FILE, FilePathCreator.create(file.getParentFile().getAbsolutePath()));
+            }
+
+            fileOutputStream = new FileOutputStream(file, APPEND);
         } catch (FileNotFoundException e) {
             Log.e(e, "File could not be opened");
-            return FilePersistenceResult.newInstance(Status.ERROR_OPENING_FILE, filePath);
+            return FilePersistenceResult.newInstance(Status.ERROR_OPENING_FILE);
+        }
+
+        return FilePersistenceResult.newInstance(Status.SUCCESS, absoluteFilePath);
+    }
+
+    private boolean createParentDirectoriesIfDoesNotExist(File outputFile) {
```

Comment:
`createParentDirectoryIfDoesNotExist()`?

Reply:
or do not exist if directories

Reply:
in general, consistency between singular and plural

Reply:
yea this is a bit weird 😆